velox/functions/sparksql/fuzzer/CMakeLists.txt

```
@@ -12,15 +12,113 @@
 # limitations under the License.
 
+# Download Spark source code and copy the Spark connect proto files to Velox
+# directory.
+set(SPARK_VERSION 3.5.1)
+set(TAR_URL
+    "https://github.com/apache/spark/archive/refs/tags/v${SPARK_VERSION}.tar.gz"
+)
+set(TAR_DEST "${CMAKE_BINARY_DIR}/spark-${SPARK_VERSION}.tar.gz")
+
+file(
+  DOWNLOAD ${TAR_URL} ${TAR_DEST}
+  STATUS status
+  LOG log)
+
+list(GET status 0 status_code)
+if(NOT status_code EQUAL 0)
+  message(FATAL_ERROR "Error downloading Spark source file: ${log}.")
+endif()
+
+execute_process(COMMAND ${CMAKE_COMMAND} -E tar xzf ${TAR_DEST}
+                WORKING_DIRECTORY ${CMAKE_BINARY_DIR})
+
+set(SOURCE_DIR
+    "${CMAKE_BINARY_DIR}/spark-3.5.1/connector/connect/common/src/main/protobuf/spark/connect"
+)
+set(DEST_DIR "${CMAKE_CURRENT_SOURCE_DIR}/proto/spark")
+file(MAKE_DIRECTORY ${DEST_DIR})
+file(COPY ${SOURCE_DIR} DESTINATION ${DEST_DIR})
+
+# Set up Spark connect.
+file(
+  GLOB PROTO_FILES
+  RELATIVE ${PROJECT_SOURCE_DIR}
+  CONFIGURE_DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/proto/spark/connect/*.proto)
+foreach(PROTO ${PROTO_FILES})
+  get_filename_component(PROTO_DIR ${PROTO} DIRECTORY)
+  get_filename_component(PROTO_NAME ${PROTO} NAME_WE)
+  set(PROTO_OUTPUT_DIR ${PROJECT_BINARY_DIR}/spark/connect)
+  list(APPEND PROTO_SRCS "${PROTO_OUTPUT_DIR}/${PROTO_NAME}.pb.cc")
+  list(APPEND PROTO_HDRS "${PROTO_OUTPUT_DIR}/${PROTO_NAME}.pb.h")
+  list(APPEND GRPC_SRCS "${PROTO_OUTPUT_DIR}/${PROTO_NAME}.grpc.pb.cc")
+  list(APPEND GRPC_HDRS "${PROTO_OUTPUT_DIR}/${PROTO_NAME}.grpc.pb.h")
+  list(APPEND PROTO_FILES_FULL
+       "${PROJECT_SOURCE_DIR}/${PROTO_DIR}/${PROTO_NAME}.proto")
+endforeach()
+
+set(PROTO_OUTPUT_FILES ${PROTO_HDRS} ${PROTO_SRCS})
+set_source_files_properties(${PROTO_OUTPUT_FILES} PROPERTIES GENERATED TRUE)
+
+set(GRPC_OUTPUT_FILES ${GRPC_HDRS} ${GRPC_SRCS})
+set_source_files_properties(${GRPC_OUTPUT_FILES} PROPERTIES GENERATED TRUE)
+
+# Ensure that the option --proto_path is not given an empty argument.
+foreach(PROTO_PATH "${PROJECT_SOURCE_DIR}/velox/functions/sparksql/fuzzer/proto"
+                   ${Protobuf_INCLUDE_DIRS})
+  list(APPEND PROTO_PATH_ARGS --proto_path=${PROTO_PATH})
+endforeach()
+
+# Generate Spark connect hearders and sources.
+add_custom_command(
+  OUTPUT ${PROTO_OUTPUT_FILES}
+  COMMAND protobuf::protoc ${PROTO_PATH_ARGS} --cpp_out ${CMAKE_BINARY_DIR}
+          ${PROTO_FILES_FULL}
+  DEPENDS protobuf::protoc
+  COMMENT "Running PROTO compiler"
+  VERBATIM)
+
+# Generate grpc headers and sources.
+add_custom_command(
+  OUTPUT ${GRPC_OUTPUT_FILES}
+  COMMAND
+    protobuf::protoc ${PROTO_PATH_ARGS} --grpc_out=${CMAKE_BINARY_DIR}
+    --plugin=protoc-gen-grpc=$<TARGET_FILE:gRPC::grpc_cpp_plugin>
+    ${PROTO_FILES_FULL}
+  DEPENDS protobuf::protoc
+  COMMENT "Running gRPC C++ protocol buffer compiler"
+  VERBATIM)
+
+set(SRCS ${PROTO_SRCS} ${GRPC_SRCS} SparkQueryRunner.cpp)
+
+add_library(velox_spark_query_runner ${SRCS})
```

```suggestion
add_library(velox_spark_query_runner ${PROTO_SRCS} ${GRPC_SRCS} SparkQueryRunner.cpp)
```
less indirection = better

If this library is supposed to be used outside of the test executable es part of velox (which it is not as a test/fuzzer utility right?) it would need to be changed to `velox_add_library` but for now it's fine.

---

Thanks for providing the information on `velox_add_library`.